protocols/identify/src/id_transport.rs

```diff
@@ -0,0 +1,208 @@
+#[derive(Debug, Clone)]
+pub struct IdentifyTransport<TTrans> {
+    transport: TTrans,
+}
+
+impl<TTrans> IdentifyTransport<TTrans> {
+    /// Creates an `IdentifyTransport` that wraps around the given transport.
+    #[inline]
+    pub fn new(transport: TTrans) -> Self {
+        IdentifyTransport {
+            transport,
+        }
+    }
+}
+
+// TODO: don't use boxes
+impl<TTrans, TMuxer> Transport for IdentifyTransport<TTrans>
+where
+    TTrans: Transport<Output = TMuxer>,
+    TMuxer: muxing::StreamMuxer + Send + Sync + 'static,      // TODO: remove unnecessary bounds
+    TMuxer::Substream: Send + Sync + 'static,      // TODO: remove unnecessary bounds
+    TMuxer::OutboundSubstream: Send + 'static,      // TODO: remove unnecessary bounds
+    TTrans::Dial: Send + Sync + 'static,
+    TTrans::Listener: Send + 'static,
+    TTrans::ListenerUpgrade: Send + 'static,
+{
+    type Output = (PeerId, TMuxer);
+    type Listener = Box<Stream<Item = (Self::ListenerUpgrade, Multiaddr), Error = IoError> + Send>;
+    type ListenerUpgrade = Box<Future<Item = Self::Output, Error = IoError> + Send>;
+    type Dial = Box<Future<Item = Self::Output, Error = IoError> + Send>;
+
+    #[inline]
+    fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+        let (listener, new_addr) = match self.transport.listen_on(addr) {
+            Ok((l, a)) => (l, a),
+            Err((inner, addr)) => {
+                let id = IdentifyTransport {
+                    transport: inner,
+                };
+                return Err((id, addr));
+            }
+        };
+
+        let listener = listener
+            .map(move |(upgrade, remote_addr)| {
+                let upgr = upgrade
+                    .and_then(move |muxer| {
+                        IdRetreiver::new(muxer, IdentifyProtocolConfig, Endpoint::Listener)
+                    });
+                (Box::new(upgr) as Box<Future<Item = _, Error = _> + Send>, remote_addr)
+            });
+
+        Ok((Box::new(listener) as Box<_>, new_addr))
+    }
+
+    #[inline]
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
+        // We dial a first time the node.
+        let dial = match self.transport.dial(addr.clone()) {
+            Ok(d) => d,
+            Err((transport, addr)) => {
+                let id = IdentifyTransport {
+                    transport,
+                };
+                return Err((id, addr));
+            }
+        };
+
+        let dial = dial.and_then(move |muxer| {
+            IdRetreiver::new(muxer, IdentifyProtocolConfig, Endpoint::Dialer)
+        });
+
+        Ok(Box::new(dial) as Box<_>)
+    }
+
+    #[inline]
+    fn nat_traversal(&self, a: &Multiaddr, b: &Multiaddr) -> Option<Multiaddr> {
+        self.transport.nat_traversal(a, b)
+    }
+}
+
+// TODO: remove unneeded bounds
+struct IdRetreiver<TMuxer>
```

Comment:
Was `id_transport.rs` supposed to be part of this PR? Regardless you probably meant `IdRetriever` here.

Reply:
I'll submit it separately.